Add support for kubeletDiskType

willie-yao opened the pull request:
<!-- If this is your first PR, welcome! Please make sure you read the [contributing guidelines](../CONTRIBUTING.md). -->

 <!-- Please label this pull request according to what type of issue you are addressing (see ../CONTRIBUTING.md) -->
**What type of PR is this?**
/kind feature

<!--
Add one of the following kinds:
/kind feature
/kind bug
/kind api-change
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind flake
-->

**What this PR does / why we need it**:
This PR adds support for [kubeletDiskType](https://learn.microsoft.com/en-us/rest/api/aks/managed-clusters/create-or-update?tabs=HTTP#kubeletdisktype) on AzureManagedMachinePools.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2912

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add support for kubeletDiskType
```

willie-yao commented:
/retest

zioproto commented:
@willie-yao as agreed during the community meeting I will start testing as soon as you remove the `WIP` in the title, thanks !

willie-yao commented:
@zioproto Removed the WIP! The feature should be working, but will need to do some additional testing/adding docs.

nojnhuh commented on azure/services/agentpools/spec.go:
```diff
@@ -278,6 +281,7 @@ func (s *AgentPoolSpec) Parameters(existing interface{}) (params interface{}, er
 			EnableAutoScaling:    s.EnableAutoScaling,
 			EnableUltraSSD:       s.EnableUltraSSD,
 			KubeletConfig:        kubeletConfig,
+			KubeletDiskType:      containerservice.KubeletDiskType(to.String(s.KubeletDiskType)),
```
Since it looks like AKS doesn't allow this to be changed, we should add a check in the webhook like:
https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/8b6f01a615a639df20644ffd0928d778b5e9fc9e/exp/api/v1beta1/azuremanagedmachinepool_webhook.go#L120-L125

nojnhuh commented on docs/book/src/topics/managedcluster.md:
```diff
@@ -361,6 +361,24 @@ spec:
 
+## AKS Node Pool KubeletDiskType configuration
+
+You can configure the `KubeletDiskType` value for each AKS node pool (`AzureManagedMachinePool`) that you define in your spec (see [here](https://learn.microsoft.com/en-us/rest/api/aks/managed-clusters/create-or-update?tabs=HTTP#kubeletdisktype) for the official AKS documentation). There are two options to choose from: `"OS"` or `"Temporary"`.
```
I tried creating a cluster with a `spec.kubeletDiskType` `Temporary` and ran into the following error in the controller logs:
```
E1215 22:15:03.423111      71 controller.go:326]  "msg"="Reconciler error" "error"="error creating AzureManagedControlPlane default/aks-29206: failed to reconcile AzureManagedControlPlane service managedcluster: failed to create resource aks-29206/aks-29206 (service: managedcluster): Code=\"PreviewFeatureNotRegistered\" Message=\"Preview feature Microsoft.ContainerService/KubeletDisk not registered.\"" "AzureManagedControlPlane"={"name":"aks-29206","namespace":"default"} "controller"="azuremanagedcontrolplane" "controllerGroup"="infrastructure.cluster.x-k8s.io" "controllerKind"="AzureManagedControlPlane" "name"="aks-29206" "namespace"="default" "reconcileID"="aef62a22-1e54-4c34-b0de-ab7ff1a73761"
```

So it might be good to drop a hint to run this before this can be used:
```
az feature register --namespace Microsoft.ContainerService --name KubeletDisk
```

willie-yao replied:
Great catch! I actually ran into the exact same issue, so I think it'll be great to add to the docs.

nojnhuh reviewed (commented):
Took this for a spin and it works great! Left a few small comments inline.

/assign

willie-yao commented on exp/api/v1beta1/azuremanagedmachinepool_types.go:
```diff
@@ -189,6 +189,11 @@ type AzureManagedMachinePoolSpec struct {
+	// KubeletDiskType specifies the kubelet disk type. Default to OS. Possible values include: 'OS', 'Temporary'
+	// +kubebuilder:validation:Enum=OS;Temporary
+	// +optional
+	KubeletDiskType *string `json:"kubeletDiskType,omitempty"`
```
I tried doing this originally, but I was unable to get the typing to work for the agentpool spec [here](https://github.com/willie-yao/cluster-api-provider-azure/blob/8b6f01a615a639df20644ffd0928d778b5e9fc9e/azure/scope/managedmachinepool.go#L182). The agent pool type [definition](https://github.com/willie-yao/cluster-api-provider-azure/blob/8b6f01a615a639df20644ffd0928d778b5e9fc9e/azure/services/agentpools/spec.go#L134) for the KubeletDiskType is unable to reference the type if I declare it as a constant inside of AzureManagedMachinePools, and vice versa.

nojnhuh replied:
What exactly did you try where you 
I think it would work to leave the field in `AgentPoolSpec` as a `*string` (since that's what the Azure API expects) and then we can convert from the custom type with something like:
```go
		KubeletDiskType:      (*string)(managedMachinePool.Spec.KubeletDiskType),
```

willie-yao replied:
I was able to fix this by converting the custom type within the toString method [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2938/commits/c5119e1177d66b951be23793bcb58cc6ec74a66c#diff-2c4585415b3fabfd1913f95eb20bc57daac00c2facd3ffa45d3bb00bb1d41b63L284)! Thanks so much for the suggestion.

willie-yao commented on docs/book/src/topics/managedcluster.md:
```diff
@@ -361,6 +361,24 @@ spec:
 
+## AKS Node Pool KubeletDiskType configuration
+
+You can configure the `KubeletDiskType` value for each AKS node pool (`AzureManagedMachinePool`) that you define in your spec (see [here](https://learn.microsoft.com/en-us/rest/api/aks/managed-clusters/create-or-update?tabs=HTTP#kubeletdisktype) for the official AKS documentation). There are two options to choose from: `"OS"` or `"Temporary"`.
```
Nice catch!

willie-yao commented:
@zioproto How did testing go for you? Let me know if you ran into anything, or if you need me to make any changes!

zioproto commented:
I confirm that after enabling manually the preview feature:

```
az feature register --namespace microsoft.ContainerService --name KubeletDisk
# wait for feature to be in state registered 
az feature list | grep KubeletDisk
# Once the feature 'KubeletDisk' is registered get the change propagated
az provider register -n microsoft.ContainerService
```
everything worked as expected.

with `az aks show -n test -g test -o json` I can read back `agentPoolProfiles.kubeletDiskType` set to `Temporary`.

This is the yaml definition I use for testing:

```
---
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: AzureManagedMachinePool
metadata:
  name: pool0
  namespace: default
spec:
  enableNodePublicIP: false
  mode: System
  osDiskSizeGB: 30
  sku: Standard_DC4s_v2
  additionalTags:
    owner: "saverio"
  kubeletDiskType: "Temporary"
  kubeletConfig:
    cpuManagerPolicy: "static"
    cpuCfsQuota: true
    cpuCfsQuotaPeriod: "110ms"
    imageGcHighThreshold: 70
    imageGcLowThreshold: 50
    topologyManagerPolicy: "best-effort"
    allowedUnsafeSysctls:
      - "net.*"
      - "kernel.msg*"
    failSwapOn: false
    containerLogMaxSizeMB: 500
    containerLogMaxFiles: 50
    podMaxPids: 2048
```

I also tested that the field is immutable.

willie-yao commented:
/retest

willie-yao commented:
/retest

nojnhuh reviewed (commented):
Looks good overall. Could you take a pass at seeing if there are any unit tests to add? I think this is similar enough to other features that specific e2e tests are probably not necessary for now.

willie-yao commented:
@nojnhuh I added the `kubeletDiskType` field to the `fakeAgentPoolSpec` inside of [agentpools_test.go](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2938/files#diff-0314c9a106f4d4aa02d8581c7e2ce20dc77dff393265a10e1f2f1a337e0232d1R58). I also added a new unit test for the managedmachinepool scope [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2938/files#diff-665bb447e3606b614418bd95ff2e38da9bce735ce0b55184ee97d64e0b8e4b2eR627)

nojnhuh reviewed (approved):
/lgtm
Thanks @willie-yao!

jackfrancis commented:
/retest

CecileRobertMichon commented on azure/services/agentpools/spec.go:
```diff
@@ -278,6 +282,7 @@ func (s *AgentPoolSpec) Parameters(existing interface{}) (params interface{}, er
 			EnableAutoScaling:    s.EnableAutoScaling,
 			EnableUltraSSD:       s.EnableUltraSSD,
 			KubeletConfig:        kubeletConfig,
+			KubeletDiskType:      containerservice.KubeletDiskType(to.String((*string)(s.KubeletDiskType))),
```
I see this wasn't added to the normalizedProfile used to calculate updates here: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2938/files#diff-2c4585415b3fabfd1913f95eb20bc57daac00c2facd3ffa45d3bb00bb1d41b63R189 

My understanding is that it's intentional because the property is immutable but I just want to confirm?

willie-yao replied:
I think the link might be pointing to the wrong place, but I don't think this was intentional. I'm not exactly familiar with the normalizedProfile. Would this field be left out since it's immutable?

CecileRobertMichon replied:
maybe @nojnhuh or @jackfrancis can help answer this

nojnhuh replied:
GitHub is having trouble rendering that link for me too, but I'm guessing it's here? https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/5b05d6e83e330b16642c121ae515851619615498/azure/services/agentpools/spec.go#L189
@CecileRobertMichon That's not the same field being added in this change, and I see that's conditionally added to `normalizedProfile` below, so that seems fine at first glance to me, albeit slightly asymmetrical with everything else: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/5b05d6e83e330b16642c121ae515851619615498/azure/services/agentpools/spec.go#L208

To your question @willie-yao, yes, if this field is immutable, then we don't need to include it in `existingProfile` or `normalizedProfile`. A diff between those two objects is what ultimately triggers a PUT request to make an update, but the webhook enforces that this field can't change.

CecileRobertMichon commented:
/lgtm
/approve

nojnhuh commented:
/lgtm

nojnhuh commented:
/lgtm
